contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -1639,3 +1645,251 @@ pltsql_replace_non_determinstic(text *src_text, text *from_text, text *to_text,
+				NextChar(t, tlen);
+				NextChar(p, plen);
+				continue;
+			}
+			else if (*p == '[')
+			{
+				/* Tsql deal with [ and ] wild character */
+				bool find_match = false, reverse_mode = false, close_bracket = false;
+				const char * prev = NULL;
+
+				NextChar(p, plen);
+				if (plen > 0 && *p == '^')
+				{
+					reverse_mode = true;
+					NextChar(p, plen);
+				}
+				while (plen > 0)
+				{
+					if (*p == ']')
+					{
+						close_bracket = true;
+						NextChar(p, plen);
+						break;
+					}
+					if (find_match)
+					{
+						NextChar(p, plen);
+						continue;
+					}
+					if (*p == '-' && prev)
+					{
+						NextChar(p, plen);
+						Assert(cid != InvalidOid);
+						if (varstr_cmp(t, pg_mblen(t), prev, pg_mblen(prev), cid) >= 0 && varstr_cmp(t, pg_mblen(t), p, pg_mblen(p), cid) <= 0)
+						{
+							find_match = true;
+						}
+						prev = NULL;
+						NextChar(p, plen);
+					}
+					else
+					{
+						int p_start_len = plen, matched_idx = 0;
+						char *p_start = p;
+						text *src_text, *substr_text;
+						prev = p;
+
+						/* find the string till the next special character */
+						while (plen > 0 && *p != ']' && *p != '-')
+						{
+							prev = p;
+							NextChar(p, plen);
+						}
+						
+						src_text = cstring_to_text_with_len(p_start, p_start_len - plen);
+						substr_text = cstring_to_text_with_len(t, pg_mblen(t));
+						
+						if (pltsql_strpos_non_determinstic(src_text, substr_text, cid, &matched_idx) && matched_idx != 0)
+						{
+							find_match = true;
+						}
+						pfree(src_text);
+						pfree(substr_text);
+					}
+				}
+
+				if (close_bracket && (find_match ^ reverse_mode)) /* found a match */
+					NextChar(t, tlen);
+				else
+				{
+					match_failed = true;
+					break;
+				}
+			}
+			else
+			{
+				char *p_start = p;
+				int  len = plen, matched_len = 0;
+
+				while (plen > 0 && *p != '[' && *p != '_' && *p != '%')
+				{
+					NextChar(p, plen);
+				}
+				if (icu_find_matched_length(t, strlen(t), p_start, len-plen, cid, &matched_len))
+				{
+					while (matched_len--)
+						NextChar(t, tlen);
+				}
+				else
+				{
+					match_failed = true;
+					break;
+				}
+			}
+		}
+
+		if (tlen > 0 && !match_failed)
+		{
+			while (tlen > 0 && *t == ' ')
+				NextChar(t, tlen);
+			if (tlen <= 0 && plen <=0)
+				return itr;
+		}
+
+		/*
+		* End of text, but perhaps not of pattern.  Match if the remaining
+		* pattern can match a zero-length string, ie, it's zero or more %'s.
+		*/
+		while (plen > 0 && *p == '%')
+			NextChar(p, plen);
+		if (tlen == 0 && plen <= 0 && !match_failed)
+			return itr;
+
+		if (start_offset)
+			input_str += pg_mblen(input_str);
+		else
+			break;
+	}
+
+	return 0;
+}
+
+PG_FUNCTION_INFO_V1(patindex_ai_collations);
+Datum
+patindex_ai_collations(PG_FUNCTION_ARGS)
+{
+	char         *pattern = text_to_cstring(PG_GETARG_TEXT_P(0));
+	char         *input_str = text_to_cstring(PG_GETARG_TEXT_P(1));
+	Oid          cid = PG_GET_COLLATION();
```

Comment:
Let's have a check whether cid is valid or not

Reply:
`tsql_lookup_collation_table_internal` & `lc_collate_is_c` both check for `if (!OidIsValid(collation))`

Reply:
done